### PR TITLE
[Discussion] Initial attempt to replace pyGSI with M2Crypto.

### PR DESCRIPTION
**Pull request by miloszz**
At the moment it's mostly proof of concept that `pyGSI` can be replaced with `M2Crypto` in reasonable time.

What’s done:
- `dirac-proxy-init` and what is most essential for it to run
- changes in `X509*` classes (cleaning out some mess)
- awful hardcoded hacks in socket/protocol layer to avoid problems with modified X509* classes

What works:
- `dirac-proxy-init`, also with `--rfc` that includes some pyGSI magic that hasn't been converted to M2Crypto

```
[LHCbDirac v8r4p7] test $ python DIRAC/FrameworkSystem/scripts/dirac-proxy-init.py --rfc
Generating proxy...
Enter Certificate password:
Added VOMS attribute /lhcb/Role=user
Uploading proxy for lhcb_user...
Proxy generated:
subject      : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=mzdybal/CN=765557/CN=Milosz Jerzy Zdybal/CN=7287075501/CN=945804218
issuer       : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=mzdybal/CN=765557/CN=Milosz Jerzy Zdybal/CN=7287075501
identity     : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=mzdybal/CN=765557/CN=Milosz Jerzy Zdybal
timeleft     : 23:53:59
DIRAC group  : lhcb_user
rfc          : True
path         : /tmp/x509up_u78251
username     : mzdybal
properties   : NormalUser
VOMS         : True
VOMS fqan    : ['/lhcb/Role=user']

Proxies uploaded:
 DN                                                                                   | Group     | Until (GMT)
 /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=mzdybal/CN=765557/CN=Milosz Jerzy Zdybal | lhcb_user | 2018/01/03 07:17
```
```
[LHCbDirac v8r4p7] test $ lhcb-proxy-info
subject      : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=mzdybal/CN=765557/CN=Milosz Jerzy Zdybal/CN=7287075501/CN=945804218
issuer       : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=mzdybal/CN=765557/CN=Milosz Jerzy Zdybal/CN=7287075501
identity     : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=mzdybal/CN=765557/CN=Milosz Jerzy Zdybal
timeleft     : 23:53:40
DIRAC group  : lhcb_user
rfc          : True
path         : /tmp/x509up_u78251
username     : mzdybal
properties   : NormalUser
VOMS         : True
VOMS fqan    : ['/lhcb/Role=user']
```

### CODE REVIEW

**Comment by chaen** [comment]
Nice ! 
I would take this opportunity to add a lot of documentation in this part of the code which has none whatsoever. Inside the code, but also as a general description, how it works and so on. 
Finally, did you make sure that the delegation works properly ?

**Comment by miloszz** [comment]
I never used any LHCb software, what's the easiest way to check delegation?

**Comment by atsareg** [comment]
Delegation is not LHCb specific. Delegation is a mechanism to transmit user identity from client to server without passing the private key through the network

**Comment by chaen** [comment]
Upload the proxy to the proxymanager, get it back, try uploading a file to a storage element, etc. All this uses delegation

File: Core/Security/X509Certificate.py [hunk 1]
```diff
@@ -252,3 +276,13 @@ def getExtensions( self ):
         value = "Cannot decode value"
       extList.append( ( sn, value ) )
     return S_OK( sorted( extList ) )
+
+  def verify( self, pkey ):
+    ret = self.__certObj.verify( pkey )
+    return S_OK( ret )
+
+  def get_subject( self ):
+    return self.getSubjectDN()['Value']
```
Comment (atsareg) [inline Core/Security/X509Certificate.py]: Never do it this way. Result return structure must always be checked for "OK" or not

Comment (miloszz) [inline Core/Security/X509Certificate.py]: OK, will fix that. I guess it's one of the hacks to allow other parts (SSLSockets, etc.) work in kind of legacy mode. I tried not to use SSL-related object calls like get_subject outside of X509Certificate class (use getSubject -> S_OK instead), but some parts of code expected pyGSI X509 objects, not Core.Security.X509Certificate.

**Comment by atsareg** [comment]
Please, make all new development PRs from now on to target the integration branch

**Comment by fstagni** [comment]
For the sake of letting travis work, please add "M2Crypto" to the DIRAC/requirements.txt file

File: Core/Security/X509Chain.py [hunk 2]
```diff
@@ -404,8 +429,9 @@ def __checkProxyDN( self, certStep, issuerStep ):
      1 = proxy match
      2 = limited proxy match
     """
-    issuerSubject = self.__certList[ issuerStep ].get_subject()
-    proxySubject = self.__certList[ certStep ].get_subject().clone()
+    return 1
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: Mmm?

**Comment by fstagni** [comment]
There's a conflicting file. Please rebase or merge with integration and push.

**Comment by miloszz** [comment]
My bad, haven't rebased before pushing changes, wrong target branch.

I marked commented lines of code for a obvious fixes, I'll make a new PR, because my repo got very messy :(

**Pull request closed without merging**